ibis/expr/rewrites.py

```diff
@@ -132,28 +119,65 @@ def backtrack(cls, value):
         ):
             yield value, distance
 
-    def dereference(self, value):
-        """Dereference a value to the target relations.
+    def _fill_substitution_mappings(self):
+        if self.subs is not None and self.ambigs is not None:
+            return
+
+        mapping = defaultdict(dict)
+
+        for rel in self.rels:
+            for field in rel.fields.values():
+                for val, distance in self.__class__.backtrack(field):
+                    mapping[val][field] = distance
+
+        subs, ambigs = {}, {}
+        for from_, to in mapping.items():
+            mindist = min(to.values())
+            minkeys = [k for k, v in to.items() if v == mindist]
+            # if all the closest fields are from the same relation, then we
+            # can safely substitute them and we pick the first one arbitrarily
+            if all(minkeys[0].relations == k.relations for k in minkeys):
+                subs[from_] = minkeys[0]
+            else:
+                ambigs[from_] = minkeys
+
+        subs.update(self.extra)
+
+        self.subs = subs
+        self.ambigs = ambigs
```

Comment:
there are no tests to check ambigs behaves as expected ATM.  would be useful to add (could be separate PR) before above idea to cache derefmap cache levels, as I think determining them correctly might be a little trickier with the caching approach